ydb/core/persqueue/partition.cpp

```diff
@@ -979,114 +944,46 @@ void TPartition::Handle(TEvPQ::TEvTxCalcPredicate::TPtr& ev, const TActorContext
 
 void TPartition::Handle(TEvPQ::TEvTxCommit::TPtr& ev, const TActorContext& ctx)
 {
-    EndTransaction(*ev->Get(), ctx);
-
-    TxInProgress = false;
-
-    ContinueProcessTxsAndUserActs(ctx);
-}
-
-void TPartition::Handle(TEvPQ::TEvTxRollback::TPtr& ev, const TActorContext& ctx)
-{
-    EndTransaction(*ev->Get(), ctx);
-
-    TxInProgress = false;
-
-    ContinueProcessTxsAndUserActs(ctx);
-}
-
-void TPartition::Handle(TEvPQ::TEvGetWriteInfoResponse::TPtr& ev, const TActorContext& ctx)
-{
-    LOG_DEBUG_S(ctx, NKikimrServices::PERSQUEUE,
-                "TEvGetWriteInfoResponse Cookie: " << ev->Get()->Cookie);
-
-    Y_ABORT_UNLESS(TxInProgress);
-
-    bool predicate = true;
-    for (auto& [k, v] : ev->Get()->SrcIdInfo) {
-        auto s = SourceManager.Get(k);
-        if (s.State == TSourceIdInfo::EState::Unknown) {
-            continue;
-        }
-
-        if (v.MinSeqNo < s.SeqNo) {
-            predicate = false;
-            break;
-        }
-    }
-
-    WriteInfoResponse = ev->Release();
-
-    if (auto* t = TryGetCurrentImmediateTransaction()) {
-        if (!WriteInfoResponse->BodyKeys.empty()) {
-            predicate = false;
+    if (PlanStep.Defined() && TxId.Defined()) {
+        if (GetStepAndTxId(*ev->Get()) < GetStepAndTxId(*PlanStep, *TxId)) {
+            ctx.Send(Tablet, MakeCommitDone(ev->Get()->Step, ev->Get()->TxId).Release());
+            return;
         }
-
-        auto tx = std::move(t->Record);
-
-        // This is a transaction for a write operation. Write operations will be added to the head of the queue
-        UserActionAndTransactionEvents.pop_front();
-        --ImmediateTxCount;
-
-        ProcessImmediateTx(tx, predicate, ctx);
-        ScheduleTransactionCompleted(tx);
-        TxInProgress = false;
-        ContinueProcessTxsAndUserActs(ctx);
-
-        return;
     }
 
-    auto& t = GetCurrentTransaction();
-    Y_ABORT_UNLESS(t.Tx);
-
-    if (WriteInfoResponse->BodyKeys.empty()) {
-        t.Predicate = predicate && BeginTransaction(*t.Tx, ctx);
+    auto txIter = TransactionsInflight.begin();
+    if (ChangeConfig) {
+        Y_ABORT_UNLESS(TransactionsInflight.size() == 1);
```

Comment:
How do you guarantee it?

Reply:
ChangeConfig is blocked if FirstEvent flag is non set. And anything else is blocked if ChangeConfig is defined.